x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java

```diff
@@ -204,48 +214,51 @@ protected void doExecute(Task task, CloseJobAction.Request request, ActionListen
+                listener.onFailure(
+                    ExceptionsHelper.conflictStatusException("cannot close job [{}] because it failed, use force close",
+                        expandedJobIds.iterator().next()));
+                return;
+            }
+            listener.onFailure(
+                ExceptionsHelper.conflictStatusException("one or more jobs have state failed, use force close"));
+            return;
+        }
 
-                    // If there are failed jobs force close is true
-                    ids.openJobIds.addAll(failedJobs);
-                    listener.onResponse(ids);
-                },
-                listener::onFailure
-        ));
+        // If there are failed jobs force close is true
+        ids.openJobIds.addAll(failedJobs);
+        listener.onResponse(ids);
     }
 
-    void checkDatafeedsHaveStopped(Collection<String> jobIds, PersistentTasksCustomMetadata tasksMetadata,
-                                   ActionListener<Boolean> listener) {
-        datafeedConfigProvider.findDatafeedsForJobIds(jobIds, ActionListener.wrap(
+    void stopDatafeedsIfNecessary(OpenAndClosingIds jobIds, boolean isForce, TimeValue timeout, PersistentTasksCustomMetadata tasksMetadata,
+                                  ActionListener<Boolean> listener) {
+        datafeedConfigProvider.findDatafeedsForJobIds(jobIds.openJobIds, ActionListener.wrap(
                 datafeedIds -> {
-                    for (String datafeedId : datafeedIds) {
-                        DatafeedState datafeedState = MlTasks.getDatafeedState(datafeedId, tasksMetadata);
-                        if (datafeedState != DatafeedState.STOPPED) {
-                            listener.onFailure(ExceptionsHelper.conflictStatusException(
-                                "cannot close job datafeed [{}] hasn't been stopped", datafeedId));
-                            return;
-                        }
+                    List<String> runningDatafeedIds = datafeedIds
+                        .stream()
+                        .filter(datafeedId -> MlTasks.getDatafeedState(datafeedId, tasksMetadata) != DatafeedState.STOPPED)
+                        .collect(Collectors.toList());
+                    if (runningDatafeedIds.isEmpty()) {
+                        listener.onResponse(false);
+                    } else {
+                        StopDatafeedAction.Request request = new StopDatafeedAction.Request(String.join(",", runningDatafeedIds));
+                        request.setForce(isForce);
+                        request.setStopTimeout(timeout);
+                        ClientHelper.executeAsyncWithOrigin(
+                            client,
+                            ClientHelper.ML_ORIGIN,
+                            StopDatafeedAction.INSTANCE,
+                            request,
+                            ActionListener.wrap(r -> listener.onResponse(r.isStopped()), listener::onFailure));
```

Comment:
Do we know if the failure here will mention the datafeeds that failed to stop?

It might be useful to indicate "Hey, we didn't close the job becuase its datafeeds were running and we couldn't stop it". This may already be the case, just double checking.

Reply:
Good point.  I'll improve the error tomorrow.